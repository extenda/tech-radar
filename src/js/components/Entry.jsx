import React from 'react';
import PropTypes from 'prop-types';
import Markdown from 'react-markdown';
import radarService from '../modules/radarService';
import Related from './entry/Related';
import History from './entry/History';
import Badge from './entry/Badge';
import Navigation from './Navigation';
import NotFound from './NotFound';

const Entry = (props) => {
  const { match } = props;
  const entry = radarService.getEntry(match.params.id);
  if (!entry) {
    return (
      <NotFound />
    );
  }
  const quadrant = radarService.model.quadrants.find(q => q.dirname === entry.quadrant.dirname);

  return (
    <React.Fragment>
      <Navigation quadrant={quadrant} entry={{ name: entry.name }} />
      <div className="container">
        <div className="row">
          <div className="entry-container">
            <h1>{entry.name}</h1>
            {entry.logo && (
              <img src={entry.logo} alt={entry.name} className="entry-logo" />
            )}
          </div>
        </div>
        <div className="row">
          <div className="badges">
            <Badge className={entry.blip.ringName} icon="map-marker" text={entry.blip.ringName} />
            <Badge icon="clock-o" text={entry.blip.since} />
            {(entry.blip.active === false) && (
              <Badge className="inactive" icon="warning" text="Inactive" />
            )}
          </div>
          <div className="twelve.columns">
            <Markdown source={entry.description} />
          </div>
          {entry.rationale && (
            <React.Fragment>
              <h2>Rationale</h2>
              <Markdown source={entry.rationale} />
            </React.Fragment>
          )}
          <Related related={entry.related} />
          <History history={entry.blip.history} />
        </div>
      </div>
    </React.Fragment>
  );
};

Entry.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
};

export default Entry;
