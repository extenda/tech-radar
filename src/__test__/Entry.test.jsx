import React from 'react';
import { shallow } from 'enzyme';
import Entry from '../js/components/Entry';

const mockEntries = {
  java: {
    name: 'Java',
    blip: {
      label: 'Java',
      quadrant: 2,
      since: 'Mar 20, 2018',
      ring: 0,
      ringName: 'Adopt',
      moved: false,
      link: '/entries/java.html',
      active: true,
    },
    description: 'Description',
    rationale: 'Rationale',
    related: [
      { file: 'php.html', name: 'PHP' },
    ],
    quadrant: { index: 2, dirname: 'dev' },
    filename: 'java.html',
  },
  lisp: {
    label: 'Lisp',
    blip: {
      quadrant: 2,
      since: 'Jan 1, 1970',
      ring: 3,
      ringName: 'Hold',
      moved: false,
      link: '/entries/lisp.html',
      active: false,
      history: [
        { date: 'Jan 1, 1970', ringName: 'Adopt' },
        { date: 'Mar 20, 2018', ringName: 'Hold' },
      ],
    },
    description: 'Description',
    rationale: 'Rationale',
    quadrant: { index: 2, dirname: 'dev' },
    filename: 'lisp.html',
  },
  php: {
    name: 'PHP',
    logo: 'https://www.php.net/images/logos/php-logo.svg',
    blip: {
      label: 'PHP',
      quadrant: 2,
      since: 'Jan 1, 2004',
      ring: 3,
      ringName: 'Hold',
      moved: false,
      link: '/entries/php.html',
      active: true,
      history: [
        { date: 'Jan 1, 2004', ringName: 'Adopt' },
        { date: 'Mar 20, 2018', ringName: 'Hold' },
      ],
    },
    description: 'Description',
    rationale: 'Rationale',
    quadrant: { index: 2, dirname: 'dev' },
    filename: 'php.html',
  },
};

jest.mock('../js/modules/radarService', () => ({
  getEntry: entry => mockEntries[entry],
}));

describe('<Entry />', () => {
  test('It renders entry with related links', () => {
    const match = { params: { id: 'java' } };
    const component = shallow(<Entry match={match} />);
    expect(component).toMatchSnapshot();
  });

  test('It renders entry with history', () => {
    const match = { params: { id: 'php' } };
    const component = shallow(<Entry match={match} />);
    expect(component).toMatchSnapshot();
  });

  test('It renders logotype if present', () => {
    const match = { params: { id: 'php' } };
    const component = shallow(<Entry match={match} />);
    expect(component.find('img')).toHaveLength(1);
    expect(component).toMatchSnapshot();
  });

  test('It renders inactive badge', () => {
    const match = { params: { id: 'lisp' } };
    const component = shallow(<Entry match={match} />);
    expect(component.find('.inactive')).toHaveLength(1);
    expect(component).toMatchSnapshot();
  });
});
