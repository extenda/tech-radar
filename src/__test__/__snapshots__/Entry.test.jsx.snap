// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Entry /> It renders entry with history 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="entry-container"
    >
      <h1>
        PHP
      </h1>
      <img
        alt="PHP"
        className="entry-logo"
        src="https://www.php.net/images/logos/php-logo.svg"
      />
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="badges"
    >
      <Badge
        className="Hold"
        icon="map-marker"
        text="Hold"
      />
      <Badge
        className=""
        icon="clock-o"
        text="Jan 1, 2004"
      />
    </div>
    <div
      className="twelve.columns"
    >
      <ReactMarkdown
        astPlugins={Array []}
        escapeHtml={true}
        plugins={Array []}
        rawSourcePos={false}
        renderers={Object {}}
        skipHtml={false}
        source="Description"
        sourcePos={false}
        transformLinkUri={[Function]}
      />
    </div>
    <h2>
      Rationale
    </h2>
    <ReactMarkdown
      astPlugins={Array []}
      escapeHtml={true}
      plugins={Array []}
      rawSourcePos={false}
      renderers={Object {}}
      skipHtml={false}
      source="Rationale"
      sourcePos={false}
      transformLinkUri={[Function]}
    />
    <Related
      related={Array []}
    />
    <History
      history={
        Array [
          Object {
            "date": "Jan 1, 2004",
            "ringName": "Adopt",
          },
          Object {
            "date": "Mar 20, 2018",
            "ringName": "Hold",
          },
        ]
      }
    />
  </div>
</div>
`;

exports[`<Entry /> It renders entry with related links 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="entry-container"
    >
      <h1>
        Java
      </h1>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="badges"
    >
      <Badge
        className="Adopt"
        icon="map-marker"
        text="Adopt"
      />
      <Badge
        className=""
        icon="clock-o"
        text="Mar 20, 2018"
      />
    </div>
    <div
      className="twelve.columns"
    >
      <ReactMarkdown
        astPlugins={Array []}
        escapeHtml={true}
        plugins={Array []}
        rawSourcePos={false}
        renderers={Object {}}
        skipHtml={false}
        source="Description"
        sourcePos={false}
        transformLinkUri={[Function]}
      />
    </div>
    <h2>
      Rationale
    </h2>
    <ReactMarkdown
      astPlugins={Array []}
      escapeHtml={true}
      plugins={Array []}
      rawSourcePos={false}
      renderers={Object {}}
      skipHtml={false}
      source="Rationale"
      sourcePos={false}
      transformLinkUri={[Function]}
    />
    <Related
      related={
        Array [
          Object {
            "file": "php.html",
            "name": "PHP",
          },
        ]
      }
    />
    <History
      history={Array []}
    />
  </div>
</div>
`;

exports[`<Entry /> It renders inactive badge 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="entry-container"
    >
      <h1 />
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="badges"
    >
      <Badge
        className="Hold"
        icon="map-marker"
        text="Hold"
      />
      <Badge
        className=""
        icon="clock-o"
        text="Jan 1, 1970"
      />
      <Badge
        className="inactive"
        icon="warning"
        text="Inactive"
      />
    </div>
    <div
      className="twelve.columns"
    >
      <ReactMarkdown
        astPlugins={Array []}
        escapeHtml={true}
        plugins={Array []}
        rawSourcePos={false}
        renderers={Object {}}
        skipHtml={false}
        source="Description"
        sourcePos={false}
        transformLinkUri={[Function]}
      />
    </div>
    <h2>
      Rationale
    </h2>
    <ReactMarkdown
      astPlugins={Array []}
      escapeHtml={true}
      plugins={Array []}
      rawSourcePos={false}
      renderers={Object {}}
      skipHtml={false}
      source="Rationale"
      sourcePos={false}
      transformLinkUri={[Function]}
    />
    <Related
      related={Array []}
    />
    <History
      history={
        Array [
          Object {
            "date": "Jan 1, 1970",
            "ringName": "Adopt",
          },
          Object {
            "date": "Mar 20, 2018",
            "ringName": "Hold",
          },
        ]
      }
    />
  </div>
</div>
`;

exports[`<Entry /> It renders logotype if present 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="entry-container"
    >
      <h1>
        PHP
      </h1>
      <img
        alt="PHP"
        className="entry-logo"
        src="https://www.php.net/images/logos/php-logo.svg"
      />
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="badges"
    >
      <Badge
        className="Hold"
        icon="map-marker"
        text="Hold"
      />
      <Badge
        className=""
        icon="clock-o"
        text="Jan 1, 2004"
      />
    </div>
    <div
      className="twelve.columns"
    >
      <ReactMarkdown
        astPlugins={Array []}
        escapeHtml={true}
        plugins={Array []}
        rawSourcePos={false}
        renderers={Object {}}
        skipHtml={false}
        source="Description"
        sourcePos={false}
        transformLinkUri={[Function]}
      />
    </div>
    <h2>
      Rationale
    </h2>
    <ReactMarkdown
      astPlugins={Array []}
      escapeHtml={true}
      plugins={Array []}
      rawSourcePos={false}
      renderers={Object {}}
      skipHtml={false}
      source="Rationale"
      sourcePos={false}
      transformLinkUri={[Function]}
    />
    <Related
      related={Array []}
    />
    <History
      history={
        Array [
          Object {
            "date": "Jan 1, 2004",
            "ringName": "Adopt",
          },
          Object {
            "date": "Mar 20, 2018",
            "ringName": "Hold",
          },
        ]
      }
    />
  </div>
</div>
`;
